internal/cmd/query/query.go

```diff
@@ -179,29 +176,119 @@ func run(ctx context.Context, opts *options) error {
 	})
 	if err != nil {
 		return err
-	} else if res == nil || len(res.Matches) == 0 {
+	} else if len(res.Matches) == 0 && len(res.Buckets.Totals) == 0 {
 		return errors.New("query returned no results")
 	}
 
-	if opts.IO.IsStdoutTTY() {
-		s := cs.Bold(opts.Query)
-		if res.SavedQueryID != "" {
-			s += fmt.Sprintf(" (saved as %s)", cs.Bold(res.SavedQueryID))
+	progStop()
+
+	pagerStop, err := opts.IO.StartPager(ctx)
+	if err != nil {
+		return err
+	}
+	defer pagerStop()
+
+	// Deal with JSON output format.
+	if opts.Format == iofmt.JSON.String() {
+		var data any
+		if len(res.Matches) > 0 {
+			data = res.Matches
+		} else if len(res.Buckets.Totals) > 0 {
+			// For ungrouped buckets we just return the aggregated total.
+			if len(res.Buckets.Totals[0].Group) == 0 {
+				data = res.Buckets.Totals[0].Aggregations[0].Value
+			} else {
+				data = res.Buckets.Totals
+			}
+		}
+		return iofmt.FormatToJSON(opts.IO.Out(), data, opts.IO.ColorEnabled())
+	}
+
+	cs := opts.IO.ColorScheme()
+
+	headerText := cs.Bold(opts.Query)
+	if res.SavedQueryID != "" {
+		headerText += fmt.Sprintf(" (saved as %s)", cs.Bold(res.SavedQueryID))
+	}
+	headerText += cs.Gray(fmt.Sprintf(" processed in %s", res.Status.ElapsedTime))
+	headerText = fmt.Sprintf("Result of query %s:\n\n", headerText)
+
+	// Deal with table output format for matches.
+	if len(res.Matches) > 0 {
+		if opts.IO.IsStdoutTTY() {
+			fmt.Fprint(opts.IO.Out(), headerText)
+		}
+
+		var enc interface {
+			Encode(any) error
+		}
+		if opts.IO.ColorEnabled() {
+			enc = jsoncolor.NewEncoder(opts.IO.Out())
+		} else {
+			enc = json.NewEncoder(opts.IO.Out())
+		}
+
+		var data any
+		for _, entry := range res.Matches {
+			switch opts.Format {
+			case iofmt.JSON.String():
+				data = entry
+			default:
+				fmt.Fprintf(opts.IO.Out(), "%s\t", cs.Gray(entry.Time.Format(time.RFC1123)))
+				data = entry.Data
+			}
+			if err = enc.Encode(data); err != nil {
+				return err
+			}
+			fmt.Fprintln(opts.IO.Out())
 		}
-		fmt.Fprintf(opts.IO.Out(), "Result of query %s:\n\n", s)
+
+		return nil
 	}
 
-	for _, entry := range res.Matches {
-		switch opts.Format {
-		case iofmt.JSON.String():
-			_ = enc.Encode(entry)
-		default:
-			fmt.Fprintf(opts.IO.Out(), "%s\t",
-				cs.Gray(entry.Time.Format(time.RFC1123)))
-			_ = enc.Encode(entry.Data)
+	// Deal with table output format for grouped results.
+
+	// If we have no groups, we just print the aggregated total.
+	if len(res.Buckets.Totals[0].Group) == 0 {
+		if opts.IO.IsStdoutTTY() {
+			fmt.Fprint(opts.IO.Out(), headerText)
 		}
+
+		if err = iofmt.FormatToJSON(opts.IO.Out(),
+			res.Buckets.Totals[0].Aggregations[0].Value, opts.IO.ColorEnabled()); err != nil {
+			return err
+		}
+
 		fmt.Fprintln(opts.IO.Out())
+		return nil
+
```

Comment:
Linting issue here, looks good other than that, however CI needs to be passed first.